### PR TITLE
fix: harden Kimi hook placeholder cleanup

### LINKED ISSUE #498: Kimi hook installer writes invalid TOML when config has hooks = []
## 环境信息

| 项目 | 信息 |
| --- | --- |
| macOS | 26.2 (Build 25C56) |
| Swift | Apple Swift 6.3.2 (swiftlang-6.3.2.1.108 clang-2100.1.1.101), target arm64-apple-macosx26.0 |
| Open Island | `/Applications/Open Island.app/Contents/MacOS/OpenIslandApp` 正在运行；`CFBundleShortVersionString` 读取为 unknown |
| Kimi CLI | `kimi, version 1.29.0` |
| Terminal | Ghostty / zsh |

## 问题描述

当 `~/.kimi/config.toml` 里已有顶层空数组配置：

```toml
hooks = []
```

Open Island 安装 Kimi hooks 后，会继续追加受管的 `[[hooks]]` array-of-tables：

```toml
hooks = []

# open-island: managed hook — do not edit
[[hooks]]
event = "SessionStart"
matcher = "startup|resume"
command = "'/Users/xiongyingchao/Library/Application Support/OpenIsland/bin/OpenIslandHooks' --source kimi"
timeout = 45
```

这会让同一个 TOML key `hooks` 同时被定义成普通数组和 array-of-tables。Kimi 启动时解析配置失败：

```text
Invalid TOML in configuration file /Users/xiongyingchao/.kimi/config.toml: Key "hooks" already exists. at line 100 col 0
```

从源码看，相关逻辑在 `Sources/OpenIslandCore/KimiHookInstaller.swift`：`installConfigTOML()` 先用 `stripManagedBlocks()` 清理旧的 Open Island 受管 `[[hooks]]` 块，再追加新的 `[[hooks]]`。但 `stripManagedBlocks()` 只处理带 marker 的 `[[hooks]]` 块或 command 匹配的旧块，没有处理已有的空 `hooks = []`。

我也搜了现有 issues，没有找到 `Kimi` / `hooks = []` / `Key "hooks" already exists` / `Invalid TOML` 的同类报告。#416 讨论了直接修改用户配置的总体问题，但这个是 Kimi TOML 兼容性的具体 bug。

## 复现步骤

1. 准备一个 `~/.kimi/config.toml`，其中包含顶层空数组：

   ```toml
   default_model = "kimi-code/kimi-for-coding"
   hooks = []
   ```

2. 通过 Open Island 安装 Kimi hooks（Settings 或 `OpenIslandSetup installKimi`）。
3. 查看 `~/.kimi/config.toml`，会看到文件中同时存在 `hooks = []` 和多个 `[[hooks]]` 块。
4. 运行 `kimi`。

## 期望行为 vs 实际行为

**期望行为**：Open Island 写入 Kimi hooks 后，`~/.kimi/config.toml` 仍然是合法 TOML，Kimi CLI 可以正常启动。

**实际行为**：配置文件变成非法 TOML，Kimi CLI 启动失败并报 `Key "hooks" already exists`。

## 可能的修复方向

- 安装 Kimi hooks 前，识别并移除空的顶层 `hooks = []`。
- 或改为 TOML-aware 的读写逻辑，避免把 `hooks` 写成两种 TOML 类型。
- 增加回归测试：已有 `hooks = []` 时安装 Kimi hooks，输出不应再包含顶层 `hooks = []`，且应保留有效的 `[[hooks]]` entries。

### PR DESCRIPTION
**Pull request by Octane0411**
## Summary

- remove Kimi's empty top-level `hooks = []` / `hooks = [ ]` placeholder before appending managed `[[hooks]]` entries
- preserve nested user-owned `hooks` arrays after TOML table headers, including headers with trailing comments
- ignore comments, single-line strings, and basic/literal multiline strings while identifying the placeholder
- add regression coverage for the standard collision, spaced arrays, commented table headers, and multiline-string content

## Root cause

Kimi can seed `config.toml` with a top-level `hooks = []`. Open Island appends `[[hooks]]` array-of-table entries, and TOML rejects the same key being defined with both shapes. The original #499 fixed the common case, but its line-based table detection could delete user data after a commented table header or inside a multiline string.

This branch preserves the original #499 commit and author attribution, then adds a separate hardening commit on the current `main`. Supersedes #499 and closes #498.

## Validation

- `git diff --check origin/main...HEAD`
- `swift build`
- production-module smoke against the built `OpenIslandCore`: standard placeholder removal, commented nested table preservation, and basic/literal multiline-string preservation — passed
- `swift test --filter KimiHooksTests` was attempted locally, but this machine's Command Line Tools installation cannot load the repository-wide `Testing` module; GitHub Actions is the authoritative test gate

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented an empty top-level `hooks = []` placeholder from lingering when managed hooks are appended.
  * Improved correctness by preserving `hooks`-like text inside comments/strings (including multiline strings) and in non-top-level sections.
  * Handled additional variations, including spaced forms (`hooks = [ ]`), CRLF line endings, and ensuring non-empty top-level hook arrays are kept.

* **Tests**
  * Expanded `KimiHookInstaller` coverage for placeholder removal/preservation and multiline/comment/formatting edge cases.

### CODE REVIEW

File: Sources/OpenIslandCore/KimiHookInstaller.swift [hunk 1]
```diff
@@ -146,6 +148,146 @@ public enum KimiHookInstaller {
...
+                if multilineString != nil {
+                    break
+                }
+                continue
+            }
+
+            switch bytes[index] {
+            case Character("#").asciiValue:
+                return String(decoding: result, as: UTF8.self)
+            case Character("\"").asciiValue, Character("'").asciiValue:
```
[the remark this answers is not in the document]
Comment (Octane0411) [reply Sources/OpenIslandCore/KimiHookInstaller.swift:229]: Fixed in `12edcc4`: consumed TOML strings now leave a non-whitespace syntax marker, so `hooks = ["user-hook"]` remains non-empty. Added a regression test that verifies the user array is preserved while managed hooks are appended.

File: Sources/OpenIslandCore/KimiHookInstaller.swift [hunk 2]
```diff
@@ -146,6 +148,146 @@ public enum KimiHookInstaller {
...
+    }
+
+    private static func isEmptyHooksArrayAssignment(_ syntax: String) -> Bool {
+        guard let equalsIndex = syntax.firstIndex(of: "=") else { return false }
+
+        let key = syntax[..<equalsIndex].trimmingCharacters(in: .whitespaces)
+        guard key == "hooks" else { return false }
+
+        let value = syntax[syntax.index(after: equalsIndex)...].trimmingCharacters(in: .whitespaces)
+        guard value.hasPrefix("["), value.hasSuffix("]") else { return false }
```
[the remark this answers is not in the document]
Comment (Octane0411) [reply Sources/OpenIslandCore/KimiHookInstaller.swift]: Fixed in `12edcc4`: placeholder matching now trims `.whitespacesAndNewlines`, including the retained `\r` from CRLF input. Added a CRLF regression test.

File: Sources/OpenIslandCore/KimiHookInstaller.swift [hunk 3]
```diff
@@ -146,6 +148,146 @@ public enum KimiHookInstaller {
...
+                } else {
+                    index += 1
+                    while index < bytes.count {
+                        if bytes[index] == quote, !honorsEscapes || !isEscaped(index, in: bytes) {
+                            index += 1
+                            break
+                        }
+                        index += 1
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (Octane0411) [reply Sources/OpenIslandCore/KimiHookInstaller.swift:246]: Fixed in `12edcc4` using the suggested marker approach for both single-line and triple-quoted strings. The new non-empty string-array regression test passes in the production-module smoke, alongside the existing multiline preservation cases.
